File: C++/2-Medium/00316._去除重复字母/00316._去除重复字母.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <stack>
using namespace std;

string removeDuplicateLetters(string s) {
    vector<int> count(256, 0); // 字符出现次数
    vector<bool> inStack(256, false); // 字符是否在栈中
    stack<char> st; // 用于构建最终结果的栈

    // 更新字符出现次数
    for (char c : s) count[c]++;
    
    for (char c : s) {
        // 减少字符出现的次数
        count[c]--;
        
        if (inStack[c]) continue; // 如果字符已在栈中，跳过
        
        // 字典序小于栈顶字符，且栈顶字符后面还会出现，则弹出栈顶字符
        while (!st.empty() && c < st.top() && count[st.top()] > 0) {
            inStack[st.top()] = false;
            st.pop();
        }
        
        // 将当前字符压入栈中
        st.push(c);
        inStack[c] = true;
    }
    
    // 构建最终结果
    string result = "";
    while (!st.empty()) {
        result = st.top() + result; // 注意这里的顺序
        st.pop();
    }
    
    return result;
}

int main() {
    string s = "cbacdcbc";
    cout << "Result: " << removeDuplicateLetters(s) << endl;
    return 0;
}

File: Python/2-Medium/00316._去除重复字母/00316._去除重复字母.py
from typing import Dict, Set

def remove_duplicate_letters(s: str) -> str:
    """
    去除字符串中的重复字母，确保每个字母只出现一次且结果的字典序最小。
    
    参数:
        s (str): 输入的字符串。
        
    返回:
        str: 去除重复字母后，字典序最小的字符串。
    
    示例:
        >>> remove_duplicate_letters("bcabc")
        'abc'
        >>> remove_duplicate_letters("cbacdcbc")
        'acdb'
    """
    # 记录每个字符出现的次数
    char_count: Dict[str, int] = {}
    for char in s:
        char_count[char] = char_count.get(char, 0) + 1
    
    # 使用栈来保存最终的字符串，保证字典序最小
    stack: list = []
    # 记录栈中已有的字符，避免重复加入
    in_stack: Set[str] = set()
    
    # 遍历字符串
    for char in s:
        # 当前字符如果未在栈中，则尝试加入栈
        if char not in in_stack:
            # 当栈不为空且当前字符字典序小于栈顶字符，且栈顶字符在后续还会出现时，弹出栈顶字符
            while stack and char < stack[-1] and char_count[stack[-1]] > 0:
                in_stack.remove(stack.pop())
            # 当前字符压栈，并标记为已入栈
            stack.append(char)
            in_stack.add(char)
        
        # 无论是否加入栈，当前字符的计数都应减1
        char_count[char] -= 1
    
    # 将栈中的字符合并成最终字符串
    return ''.join(stack)

# 运行代码示例
print(remove_duplicate_letters("bcabc"))  # 期望输出: "abc"
print(remove_duplicate_letters("cbacdcbc"))  # 期望输出: "acdb"
